src/lib/linux-api/src/signal.rs

```diff
@@ -0,0 +1,1197 @@
+}
+
+/// Codes for SIGSEGV
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeSegv {
+    SEGV_MAPERR = bindings::LINUX_SEGV_MAPERR as i32,
+    SEGV_ACCERR = bindings::LINUX_SEGV_ACCERR as i32,
+    SEGV_BNDERR = bindings::LINUX_SEGV_BNDERR as i32,
+    SEGV_PKUERR = bindings::LINUX_SEGV_PKUERR as i32,
+    SEGV_ACCADI = bindings::LINUX_SEGV_ACCADI as i32,
+    SEGV_ADIDERR = bindings::LINUX_SEGV_ADIDERR as i32,
+    SEGV_ADIPERR = bindings::LINUX_SEGV_ADIPERR as i32,
+    SEGV_MTEAERR = bindings::LINUX_SEGV_MTEAERR as i32,
+    SEGV_MTESERR = bindings::LINUX_SEGV_MTESERR as i32,
+}
+
+/// Codes for SIGBUS
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeBus {
+    BUS_ADRALN = bindings::LINUX_BUS_ADRALN as i32,
+    BUS_ADRERR = bindings::LINUX_BUS_ADRERR as i32,
+    BUS_OBJERR = bindings::LINUX_BUS_OBJERR as i32,
+    BUS_MCEERR_AR = bindings::LINUX_BUS_MCEERR_AR as i32,
+    BUS_MCEERR_AO = bindings::LINUX_BUS_MCEERR_AO as i32,
+}
+
+/// Codes for SIGTRAP
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeTrap {
+    TRAP_BRKPT = bindings::LINUX_TRAP_BRKPT as i32,
+    TRAP_TRACE = bindings::LINUX_TRAP_TRACE as i32,
+    TRAP_BRANCH = bindings::LINUX_TRAP_BRANCH as i32,
+    TRAP_HWBKPT = bindings::LINUX_TRAP_HWBKPT as i32,
+    TRAP_UNK = bindings::LINUX_TRAP_UNK as i32,
+    TRAP_PERF = bindings::LINUX_TRAP_PERF as i32,
+}
+
+/// Codes for SIGIO/SIGPOLL
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodePoll {
+    POLL_IN = bindings::LINUX_POLL_IN as i32,
+    POLL_OUT = bindings::LINUX_POLL_OUT as i32,
+    POLL_MSG = bindings::LINUX_POLL_MSG as i32,
+    POLL_ERR = bindings::LINUX_POLL_ERR as i32,
+    POLL_PRI = bindings::LINUX_POLL_PRI as i32,
+    POLL_HUP = bindings::LINUX_POLL_HUP as i32,
+}
+
+/// Codes for SIGSYS
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeSys {
+    SYS_SECCOMP = bindings::LINUX_SYS_SECCOMP as i32,
+}
+
+#[allow(non_camel_case_types)]
+pub type linux_siginfo_t = bindings::linux_siginfo_t;
+
+pub type SigInfoDetailsKill = bindings::linux__sifields__bindgen_ty_1;
+pub type SigInfoDetailsTimer = bindings::linux__sifields__bindgen_ty_2;
+pub type SigInfoDetailsRt = bindings::linux__sifields__bindgen_ty_3;
+pub type SigInfoDetailsSigChld = bindings::linux__sifields__bindgen_ty_4;
+pub type SigInfoDetailsSigFault = bindings::linux__sifields__bindgen_ty_5;
+pub type SigInfoDetailsSigPoll = bindings::linux__sifields__bindgen_ty_6;
+pub type SigInfoDetailsSigSys = bindings::linux__sifields__bindgen_ty_7;
```

Comment:
This seems like it could be brittle if the kernel adds another field or adds a padding field somewhere in the middle. Maybe these types should be hard-coded here, and then if the bindings were to change and the types didn't match, it would cause a compile error?

Reply:
A bit, but that'll only happen when we explicitly change what version of the Linux source we're building from. It should be fairly clear what caused the breakage and how to sort it out, e.g. looking at the diff of the bindings. 

When constructing a `linux_siginfo_t` we have to directly use the bindgen types. We could create our own same-shaped types with conversion functions to use in public output for this module, but IMO it's not worth the boilerplate since we'd still have to resolve the breakage inside this module if these type names change.

Added a block-comment here to help, and added an internal constructor in SigInfo to also avoid naming the generated union field names more than we have to.

Reply:
> A bit, but that'll only happen when we explicitly change what version of the Linux source we're building from. It should be fairly clear what caused the breakage and how to sort it out, e.g. looking at the diff of the bindings.

I'm not really thinking of a compile breakage, but if for example an extra padding field was added between the fields of type `linux__sifields__bindgen_ty_6` and `linux__sifields__bindgen_ty_7`. If `linux__sifields__bindgen_ty_7` was originally a `u32` and the new `linux__sifields__bindgen_ty_7` padding field is a `u16`, `SigInfoDetailsSigSys` would change from a `u32` to a `u16` and it might break things in a way that still compiles but is incorrect, which might be difficult to detect.

But I didn't realize that all of these types are structs, so maybe this wouldn't happen then (a padding field wouldn't be a struct, so would probably be a compile error if the type changed from a struct to an integer).

Reply:
> If linux__sifields__bindgen_ty_7 was originally a u32

I don't think bindgen would create the type `linux__sifields__bindgen_ty_7` in that case; it would just use `u32`. These type names are generated for anonymous union and anonymous struct types.